File: Stack/Java/Minimum String Length After Removing Substrings.java
// Problem link: https://leetcode.com/problems/minimum-string-length-after-removing-substrings/description
// Time: O(n)
// Space: O(n)
class Solution {
    public int minLength(String s) {
        Stack<Character> stack = new Stack<>();
        stack.push('0');

        for(char chr : s.toCharArray()) {
            if(chr == 'D' && stack.peek() == 'C') {
                stack.pop();
            }
            else if(chr == 'B' && stack.peek() == 'A') {
                stack.pop();
            }
            else {
                stack.push(chr);
            }
        }

        return stack.size() - 1;
    }
}

File: Stack/Python/Minimum String Length After Removing Substrings.py
# Problem link: https://leetcode.com/problems/minimum-string-length-after-removing-substrings/description
# Time: O(n)
# Space: O(n)
class Solution:
    def minLength(self, s: str) -> int:
        stack = ['0']

        for char in s:
            if char == 'D' and stack[-1] == 'C':
                stack.pop()
            elif char == 'B' and stack[-1] == 'A':
                stack.pop()
            else:
                stack.append(char)
        
        return len(stack) - 1
